sys/kern/tmpfs.c

```diff
@@ -536,22 +673,51 @@ static int tmpfs_get_vnode(mount_t *mp, tmpfs_node_t *tfn, vnode_t **vp) {
   return 0;
 }
 
+static int tmpfs_dir_extend(tmpfs_node_t *tfn) {
+  int error;
+  tmpfs_mount_t *tfm = TMPFS_ROOT_OF(tfn->tfn_vnode->v_mount);
+
+  if ((error = tmpfs_resize(tfm, tfn, tfn->tfn_size + BLOCK_SIZE)))
+    return error;
+
+  blkptr_t blk = *tmpfs_get_blk(tfn, BLKNO(tfn->tfn_size) - 1);
+
+  size_t ndirent = BLOCK_SIZE / sizeof(tmpfs_dirent_t);
+  for (size_t i = 0; i < ndirent; i++) {
+    tmpfs_dirent_t *de =
+      (tmpfs_dirent_t *)(((uint8_t *)blk) + i * sizeof(tmpfs_dirent_t));
+    TAILQ_INSERT_TAIL(&tfn->tfn_dir.fdirents, de, tfd_entries);
+  }
+
+  return 0;
+}
+
 /*
  * tmpfs_alloc_dirent: allocate a new directory entry.
  */
-static int tmpfs_alloc_dirent(const char *name, size_t namelen,
-                              tmpfs_dirent_t **dep) {
+static int tmpfs_alloc_dirent(tmpfs_node_t *tfn, const char *name,
+                              size_t namelen, tmpfs_dirent_t **dep) {
+  int error = 0;
+
   if (namelen + 1 > TMPFS_NAME_MAX)
     return ENAMETOOLONG;
 
-  tmpfs_dirent_t *dirent = kmalloc(M_TMPFS, sizeof(tmpfs_dirent_t), M_ZERO);
+  if (TAILQ_EMPTY(&tfn->tfn_dir.fdirents)) {
+    if ((error = tmpfs_dir_extend(tfn)))
+      return error;
+  }
+
+  tmpfs_dirent_t *dirent = TAILQ_FIRST(&tfn->tfn_dir.fdirents);
+  TAILQ_REMOVE(&tfn->tfn_dir.fdirents, dirent, tfd_entries);
+  bzero(dirent, sizeof(tmpfs_dirent_t));
+
   dirent->tfd_node = NULL;
   dirent->tfd_namelen = namelen;
   memcpy(dirent->tfd_name, name, namelen);
   dirent->tfd_name[namelen] = 0;
 
   *dep = dirent;
-  return 0;
+  return error;
```

Comment:
If I understand correctly here we always return 0.
Maybe it will be better to write it explicite?

Reply:
Makes sense to me.